keps/sig-network/2136-np-version-status/README.md

```diff
@@ -0,0 +1,789 @@
+higher than the highest version the plugin knows about, then the plugin
+has no way of telling whether it can implement the policy or not, and no
+way of knowing for sure if it's correctly interpreting even the parts
+that it can see. (For example, if a network plugin that only implemented
+version "1.9" received a policy with a combined `podSelector` and
+`namespaceSelector`, it might misinterpret it by reading the
+`namespaceSelector` but not the `podSelector`, and thereby allowing
+access to/from more pods than was intended.) Thus, it is best for it to
+not even try.
+
+On the other hand, if the plugin knows about a feature but doesn't
+implement it, they may still be able to implement the other parts of the
+policy safely. (For example, a plugin that does not implement egress
+policies at all might choose to still implement the ingress side of a
+mixed ingress/egress policy.)
+
+#### The "Enforcing" Condition
+
+The `"Enforcing"` condition indicates whether the plugin is currently
+enforcing the policy described by a NetworkPolicy.
+
+- If the plugin is intentionally not enforcing the policy (eg, because
+  it depends on unsupported features), it should set a `{ type: "Enforcing",
+  status: "False" }` condition, with an appropriate `message` and the
+  same `reason` as the `"Supported"` condition.
+
+- A plugin which is implementing a policy but is not able to determine
+  when the policy is fully in effect should set a status condition with
+  `{ type: "Enforcing", status: "Unknown" }` and an appropriate `reason`
+  and `message`. This indicates to the user that the enforcing status is
+  not known and is not going to become known.
+
+- A plugin which is able to determine when a NetworkPolicy is fully in
+  effect _may_ set an initial status condition with `{ type:
+  "Enforcing", status: "False", reason: "Pending" }`, and an appropriate
+  `message`. In particular, it should do this if it thinks there may be
+  a noticeable delay before it is able to set the condition to `"True"`.
+
+- A plugin which is able to determine when a NetworkPolicy is fully in
+  effect should set a status condition with `{ type: "Enforcing",
+  status: "True" }`, and an appropriate `reason` and `message`, once it
+  knows the NetworkPolicy is fully in effect.
+
+  "Fully in effect" means that for any pod which ought to be subject
+  to the policy, either (a) the pod was subject to the policy at the time
+  when the `"Enforcing"` condition became `"True"`, or (b) the pod was not
+  yet "Ready" at the time the `"Enforcing"` condition became `"True"`,
+  but will not become "Ready" until it is subject to the policy.
+
+```
+<<[UNRESOLVED pod-readiness ]>>
+
+I kind of just required that plugins that implement "Enforcing" also
+properly set up pod ReadinessConditions for NetworkPolicy...
+
+<<[/UNRESOLVED]>>
+```
+
+#### The "Problem" Condition
+
+A `"Problem"` condition with a status of `"True"` indicates some problem
```

Comment:
two qs: 
- so, this is null if unsupported? 
- whats the diff between "not sure cause there might be some missing whitelistings that i havent done bc i cant connect to the apiserver to get updated pod info for new pods for this rule" and "iptables is borked so i cant apply any allow rules " ---- are those **both** considered `Problems`? 

i like this field just wanna probe the boundarys of it

Reply:
Hm... I was only thinking about the case of "static" problems with the policy that are there before you even try to enforce it (eg, the ambiguous CIDR problem), not "dynamic" problems with the actual enforcement of the policy.

Especially, if there is a plugin-wide problem like "iptables is borked" then it doesn't seem like that should be indicated on specific NetworkPolicy objects.

I'm not sure about the naming of `"Problem"` either. I originally wrote it as `"Warning"` but `{ type: Warning, status: True }` seems silly. Then I thought maybe it should just be a `Warning string` field rather than a condition, but then you lose the transition time, etc, which might be useful.

Reply:
hmm... spitballing
- Validated (true/false)
- Applied (true/false)
- Applicable (true/false)
- Accepted (true/false)

Reply:
I think you're still misinterpreting my intent. This is for warnings, not errors. It means the policy _is_ applied/applicable/accepted, but the plugin feels the need to issue some sort of disclaimer about it. eg, "here's how I interpreted that ambiguous CIDR string". Or eg, as described in https://kubernetes.io/docs/concepts/services-networking/network-policies/#behavior-of-to-and-from-if you use `ipBlock` on cluster ingress traffic, it's not guaranteed whether you're matching against the actual source IP, or the IP of the load balancer, or something else. A plugin could add some sort of clarification about that. (Although maybe that would work better in a "global status".)

Reply:
Yeah, **warning** sounds like the right term for it , even though it sounds silly

Reply:
+1 for warning.

I think that even silly, a Warning might not report 'a Problem' but 'a possible Problem/misbehavior' (like the CIDR case)